File: 1283-reformat-date/reformat-date.dart
enum Month {
  January('01'),
  February('02'),
  March('03'),
  April('04'),
  May('05'),
  June('06'),
  July('07'),
  August('08'),
  September('09'),
  October('10'),
  November('11'),
  December('12'),
  ;

  final String number;

  const Month(this.number);

  factory Month.fromString(String string) {
    switch (string) {
      case 'Jan':
        return Month.January;
      case 'Feb':
        return Month.February;
      case 'Mar':
        return Month.March;
      case 'Apr':
        return Month.April;
      case 'May':
        return Month.May;
      case 'Jun':
        return Month.June;
      case 'Jul':
        return Month.July;
      case 'Aug':
        return Month.August;
      case 'Sep':
        return Month.September;
      case 'Oct':
        return Month.October;
      case 'Nov':
        return Month.November;
      case 'Dec':
        return Month.December;
      default:
        throw ArgumentError('Invalid month string: $string');
    }
  }
}

class Solution {
  String reformatDate(String date) {
    final parts = date.split(' ');
    final day = parts[0];
    final month = parts[1];
    final year = parts[2];
    final regExp = RegExp(r'\d+');
    final match = regExp.firstMatch(day);
    final formattedDay =
        match!.group(0)!.length == 2 ? match.group(0) : '0${match.group(0)}';
    return '${year}-${Month.fromString(month).number}-${formattedDay}';
  }
}

File: 1283-reformat-date/reformat-date.rs
use std::collections::HashMap;

impl Solution {
    pub fn reformat_date(date: String) -> String {
        let month_map: HashMap<&str, &str> = [
            ("Jan", "01"),
            ("Feb", "02"),
            ("Mar", "03"),
            ("Apr", "04"),
            ("May", "05"),
            ("Jun", "06"),
            ("Jul", "07"),
            ("Aug", "08"),
            ("Sep", "09"),
            ("Oct", "10"),
            ("Nov", "11"),
            ("Dec", "12"),
        ].iter().cloned().collect();

        let parts: Vec<&str> = date.split_whitespace().collect();
        if parts.len() < 3 {
            return "Invalid date format".to_string();
        }

        let day = parts[0];
        let month = parts[1];
        let year = parts[2];
        let day_digits = day.chars().filter(|c| c.is_digit(10)).collect::<String>();
        let formatted_day = if day_digits.len() == 1 {
            format!("0{}", day_digits)
        } else {
            day_digits
        };

        let month_number = match month_map.get(month) {
            Some(&number) => number,
            None => return "Invalid month string".to_string(),
        };

        format!("{}-{}-{}", year, month_number, formatted_day)
    }
}
